### PR TITLE
8337658: ZGC: Move soft reference handling out of the driver loop function

### PR DESCRIPTION
**Pull request by stefank**
The ZDriver code is written to be neat and have a clear outline. The soft reference handling distracts when reading this code. I propose that we hide it a bit.

I've also clarified in comments and names that the code is dealing with clearing of *all* references.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8337658](https://bugs.openjdk.org/browse/JDK-8337658): ZGC: Move soft reference handling out of the driver loop function (**Enhancement** - P4)

### Reviewers
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - **Reviewer**) Review applies to [5a69a95e](https://git.openjdk.org/jdk/pull/20418/files/5a69a95e479896b9ea0bc625ce5f8437282de437)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) Review applies to [5a69a95e](https://git.openjdk.org/jdk/pull/20418/files/5a69a95e479896b9ea0bc625ce5f8437282de437)
 * [Guoxiong Li](https://openjdk.org/census#gli) (@lgxbslgx - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20418/head:pull/20418` \
`$ git checkout pull/20418`

Update a local copy of the PR: \
`$ git checkout pull/20418` \
`$ git pull https://git.openjdk.org/jdk.git pull/20418/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20418`

View PR using the GUI difftool: \
`$ git pr show -t 20418`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20418.diff">https://git.openjdk.org/jdk/pull/20418.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20418#issuecomment-2262908912)

### CODE REVIEW

**Comment by jsikstro** [comment]
I think this change is good and agree that `ZDriverMajor::run_thread()` becomes easier to read. Since the policy is now read and set in the construction of the `ZDriverScopeMajor`, a new getter is needed from `ZGenerationOld` and in turn `ZReferenceProcessor` to retrieve the policy for the gc request. The naming clarifications seem appropriate.

**Review by xmas92 — APPROVED** [review]
lgtm.

**Review by fisk — APPROVED** [review]
This code sure has moved around a lot! Looks good.

File: src/hotspot/share/gc/z/zReferenceProcessor.hpp [hunk 1]
```diff
@@ -69,7 +69,9 @@ class ZReferenceProcessor : public ReferenceDiscoverer {
 public:
   ZReferenceProcessor(ZWorkers* workers);
 
-  void set_soft_reference_policy(bool clear);
+  void set_soft_reference_policy(bool clear_all_soft_references);
+  bool uses_clear_all_soft_reference_policy() const;
```
Comment (lgxbslgx) [inline src/hotspot/share/gc/z/zReferenceProcessor.hpp:73]: The method name should be `uses_clear_all_soft_references_policy`. Please note the letter `s` after the `reference`.

**Review by lgxbslgx — CHANGES_REQUESTED** [review]
I don't know whether `_uses_clear_all_soft_references_policy` is a good name. I may be confused by the name if I read the code first time.

One nit shown below.

**Comment by stefank** [comment]
> I don't know whether _uses_clear_all_soft_references_policy is a good name. I may be confused by the name if I read the code first time.

Could you elaborate on what the problem with the name is?

**Comment by lgxbslgx** [comment]
> > I don't know whether _uses_clear_all_soft_references_policy is a good name. I may be confused by the name if I read the code first time.
> 
> Could you elaborate on what the problem with the name is?

The two verbs `use` and `clear` are put together, which is curious at first glance. When we dive into the concrete code, we can know the `clear_all` is actually a adjective to describe the `soft_references_policy`. So I don't know whether it would confuse the newbies.

Another problem is whether we should use the third person singular form (use `uses` or `use` in `_uses_clear_all_soft_references_policy` here). I search the code in current HotSpot and both forms are used now. So I am OK with the word `uses`. Maybe we should unify the name rule in the future.

And, my problem may be caused by my poor english, so I will agree with your opinion if you approve this name after reading my comment above.

**Comment by stefank** [comment]
I fixed the inconsistency between `soft_reference_policy` and `soft_references_policy`.

> The two verbs use and clear are put together, which is curious at first glance. When I dive into the concrete code, I can know the clear_all is actually a adjective to describe the soft_references_policy. So I don't know whether it would confuse the newbies.

Thanks for the explanation. Yes, the intention is to read this as "uses 'clear all' 'soft reference policy'". I'm going to keep that name, unless someone manages to come up with a more spot-on name that is easy to understand.

> Another problem is whether we should use the third person singular form (use uses or use in _uses_clear_all_soft_references_policy here). I search the code in current HotSpot and both forms are used now. So I am OK with the word uses. Maybe we should unify the name rule in the future.

To me "use" sounds like a command to do something, and not a property. An alternative could be to change it to "is_using_clear_all_soft_reference_policy". Would that be better?

**Review by lgxbslgx — APPROVED** [review]
Looks good.

**Comment by lgxbslgx** [comment]
> To me "use" sounds like a command to do something, and not a property. An alternative could be to change it to "is_using_clear_all_soft_reference_policy". Would that be better?

These names seem similar. I approved the current patch just now.

**Comment by stefank** [comment]
Thanks for the reviews!
/integrate

**Pull request closed without merging**